base/server/src/main/java/com/netscape/cms/password/PasswordChecker.java

```diff
@@ -27,14 +35,124 @@
+        this.minPunctuationChar = minPunctuationChar;
+        this.substringMatch = substringMatch;
+        this.maxRepeatedChar = maxRepeatedChar;
+        this.cracklibCheck = cracklibCheck;
+    }
+
+    public PasswordQuality checkPassword(String password) throws EPasswordCheckException{
+        if (password == null || password.length() == 0) {
+            return PasswordQuality.CMS_PASSWORD_EMPTY_PASSWORD;
+        }
+
+        if (password.length() < minSize) {
+            return PasswordQuality.CMS_PASSWORD_INVALID_LEN;
+        }
+
+        String basePassword = null;
+        if (minUpperLetter + minLowerLetter + minNumber + minPunctuationChar > 0) {
+            StringBuilder base = new StringBuilder(password.replaceAll("^\\p{Upper}*", ""));
+            base.reverse();
+            basePassword = base.toString().replaceAll("^\\d*", "");
+        }
+        if (minUpperLetter > 0 && countLetters(basePassword, "\\p{Upper}") < minUpperLetter) {
+            return PasswordQuality.CMS_PASSWORD_MISSING_UPPER_CASE;
+        }
+        if (minLowerLetter > 0 && countLetters(basePassword, "\\p{Lower}") < minLowerLetter) {
+            return PasswordQuality.CMS_PASSWORD_MISSING_LOWER_CASE;
+        }
+        if (minNumber > 0 && countLetters(basePassword, "\\d") < minNumber) {
+            return PasswordQuality.CMS_PASSWORD_MISSING_NUMERIC;
+        }
+        if (minPunctuationChar > 0 && countLetters(basePassword, "\\p{Punct}") < minPunctuationChar) {
+            return PasswordQuality.CMS_PASSWORD_MISSING_PUNCTUATION;
+        }
+
+
+        if (substringMatch > 0) {
+            for (int i = 0; i< password.length() - (substringMatch * 2); i++) {
+                String seq = password.substring(i, i + substringMatch);
+                String invSeq = new StringBuilder(seq).reverse().toString();
+                if(password.indexOf(seq, i + substringMatch) > 0 || password.indexOf(invSeq, i + substringMatch) > 0){
+                    return PasswordQuality.CMS_PASSWORD_SEQUENCE;
+                }
+            }
+        }
+
+        if (maxRepeatedChar > 0) {
+            int mostRepeatedChar = password.chars().mapToObj(x -> (char) x)
+                    .collect(Collectors.groupingBy(x -> x, Collectors.counting()))
+                    .entrySet().stream().max(Map.Entry.comparingByValue())
+                    .get().getValue().intValue();
+            if (maxRepeatedChar < mostRepeatedChar) {
+                return PasswordQuality.CMS_PASSWORD_REPEATED_CHAR;
+            }
+        }
+
+        if (cracklibCheck) {
+            try {
+                Process crack = new ProcessBuilder("/usr/sbin/cracklib-check").start();
+                BufferedWriter crackIn = crack.outputWriter();
+                BufferedReader crackOut = crack.inputReader();
+                crackIn.write(password);
+                crackIn.close();
+
+                String crackResult = crackOut.readLine().substring(password.length() + 2);
```

Comment:
does cracklib check for substrings of the password?  If not, perhaps you could do some minimal processing first to extract consecutive string of characters from password to pass into cracklib.
We can do more complex checks in phase 2.

Reply:
It does some substring check but not sure how much it is effective but get consecutive characters does not work then I replace a character with a symbol. The word matches to dictionary if taken entirely. I think this extraction should go to phase 2.